**Author**: Moch Nabil Farras Dhiya (10120034)

**E-mail**: nabilfarras923@gmail.com

-------------------

**Disclaimer**: The **dataset** used in this analysis is a public dataset retrieved from [Customer Personality Analysis - Kaggle](https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis?datasetId=1546318&sortBy=voteCount).

# Background

## Attributes

**People**

*   ID: Customer's unique identifier
*   Year_Birth: Customer's birth year
*   Education: Customer's education level
*   Marital_Status: Customer's marital status
*   Income: Customer's yearly household income
*   Kidhome: Number of children in customer's household
*   Teenhome: Number of teenagers in customer's household
*   Dt_Customer: Date of customer's enrollment with the company
*   Recency: Number of days since customer's last purchase
*   Complain: 1 if the customer complained in the last 2 years, 0 otherwise

**Products**

*   MntWines: Amount spent on wine in last 2 years
*   MntFruits: Amount spent on fruits in last 2 years
*   MntMeatProducts: Amount spent on meat in last 2 years
*   MntFishProducts: Amount spent on fish in last 2 years
*   MntSweetProducts: Amount spent on sweets in last 2 years
*   MntGoldProds: Amount spent on gold in last 2 years

**Promotion**


*   NumDealsPurchases: Number of purchases made with a discount
*   AcceptedCmp1: 1 if customer accepted the offer in the 1st campaign, 0 otherwise
*   AcceptedCmp2: 1 if customer accepted the offer in the 2nd campaign, 0 otherwise
*   AcceptedCmp3: 1 if customer accepted the offer in the 3rd campaign, 0 otherwise
*   AcceptedCmp4: 1 if customer accepted the offer in the 4th campaign, 0 otherwise
*   AcceptedCmp5: 1 if customer accepted the offer in the 5th campaign, 0 otherwise
*   Response: 1 if customer accepted the offer in the last campaign, 0 otherwise

**Place**

*   NumWebPurchases: Number of purchases made through the company’s website
*   NumCatalogPurchases: Number of purchases made using a catalogue
*   NumWebVisitsMonth: Number of visits to company’s website in the last month
*   NumStorePurchases: Number of purchases made directly in stores

## Goals

Perform clustering to summarize customer segments.

# Connect to Google Drive

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [2]:
%cd /content/gdrive/My Drive/Portfolio/Data Science/Python/Customer Segmentation/CSV

/content/gdrive/My Drive/Portfolio/Data Science/Python/Customer Segmentation/CSV


# Import Packages

In [3]:
import pandas as pd
import numpy as np
import datetime as dt

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style("darkgrid")

from statistics import mean
from scipy.stats import skew

from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import davies_bouldin_score, silhouette_score, calinski_harabasz_score
from sklearn.metrics.pairwise import euclidean_distances
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans, AgglomerativeClustering

from yellowbrick.cluster import KElbowVisualizer
from collections import defaultdict

# Import Data

In [4]:
# data = pd.read_csv("../Data/final_data.csv", index_col = 0)

# rfm = pd.read_csv("../Data/RFM/rfm_analysis.csv", index_col = 0)

data = pd.read_csv("final_data.csv", index_col = 0)

rfm = pd.read_csv("rfm_analysis.csv", index_col = 0)

# Final Segmentation

In [5]:
data = pd.merge(data[['ID', 'Marital_Status', 'Education', 'Income_Class', 'Age_Class', 'Kidhome', 'Teenhome']], rfm, how = 'inner', on = 'ID')
data

ID  Marital_Status  Education  Income_Class  Age_Class  Kidhome  \
0      5524               0          1             2          3        0   
1      2174               0          1             1          3        1   
2      4141               1          1             3          2        0   
3      6182               1          1             0          0        1   
4      5324               1          3             2          1        1   
...     ...             ...        ...           ...        ...      ...   
2207  10870               1          1             2          2        0   
2208   4001               1          3             2          3        2   
2209   7270               0          1             2          1        0   
2210   8235               1          2             3          3        0   
2211   9405               1          3             2          3        1   

      Teenhome  Recency_Cluster  Frequency_Cluster  Wines_Cluster  \
0            0                1                  2              1   
1            1                1                  0              0   
2            0                0                  1              1   
3            0                0                  0              0   
4            0                2                  1              0   
...        ...              ...                ...            ...   
2207         1                1                  1              1   
2208         1                1                  2              1   
2209         0                2                  1              2   
2210         1                0                  1              1   
2211         1                1                  1              0   

      Fruits_Cluster  Meat_Cluster  Fish_Cluster  Sweets_Cluster  Gold_Cluster  
0                  1             2             2               1             1  
1                  0             0             0               0             0  
2                  1             0             1               0             0  
3                  0             0             0               0             0  
4                  1             0             1               0             0  
...              ...           ...           ...             ...           ...  
2207               1             0             0               2             2  
2208               0             0             0               0             0  
2209               1             1             0               0             0  
2210               0             1             1               0             1  
2211               0             0             0               0             0  

[2212 rows x 15 columns]

In [6]:
min_customer = 10

## Functions

In [7]:
# Potential Wines Customer
def get_potential_customer(df, product_cluster, cluster):
  temp = df.loc[df[product_cluster] == cluster][['Marital_Status', 'Education', 'Income_Class', 'Age_Class', 'Kidhome', 'Teenhome']] \
          .groupby(['Marital_Status', 'Education', 
                    'Income_Class', 'Age_Class', 
                    'Kidhome', 'Teenhome']) \
          .value_counts().reset_index()
  temp = temp.rename(columns = {0: 'Potential_Customers'})

  return temp

In [8]:
def get_potential_percentage(df, temp, min_customer):
  re = df[['Marital_Status', 'Education',  'Income_Class', 'Age_Class', 'Kidhome', 'Teenhome']] \
            .groupby(['Marital_Status', 'Education', 
                      'Income_Class', 'Age_Class', 
                      'Kidhome', 'Teenhome']) \
            .value_counts().reset_index()

  re = re.rename(columns = {0: 'Customers'})
  re = pd.merge(re, temp, how = 'inner', on = ['Marital_Status', 'Education', 
                                                    'Income_Class', 'Age_Class',
                                                    'Kidhome', 'Teenhome'])

  # Only locate characteristics which have at least 10 customers
  re = re.loc[re['Customers'] >= min_customer]
  re['Potential_Percentage'] = 100 * re['Potential_Customers'] / re['Customers']
  re = re.sort_values(by = 'Potential_Percentage', ascending = False).reset_index(drop = True)
  
  return re

In [9]:
def display_rfm(df, product_cluster):
  df = df[['Recency_Cluster', 'Frequency_Cluster', product_cluster]] \
            .groupby(['Recency_Cluster', 'Frequency_Cluster', product_cluster]) \
            .value_counts().reset_index()
  df = df.rename(columns = {0: 'value'})

  # Recency and Frequency Percentage
  recency_data = round(100 * df.groupby('Recency_Cluster')['value'].sum() / \
                       df['value'].sum(), 3)
  frequency_data = round(100 * df.groupby('Frequency_Cluster')['value'].sum() / \
                         df['value'].sum(), 3)
  monetary_data = round(100 * df.groupby(product_cluster)['value'].sum() / \
                        df['value'].sum(), 3)

  display(recency_data)
  display(frequency_data)
  display(monetary_data)

In [10]:
def get_concern(df, product_data, mask):
  concern_data = df[df['ID'].isin(mask)][['Marital_Status', 'Education', 
                                          'Income_Class', 'Age_Class', 
                                          'Kidhome', 'Teenhome']] \
                      .groupby(['Marital_Status', 'Education', 
                                'Income_Class', 'Age_Class', 
                                'Kidhome', 'Teenhome']) \
                      .value_counts().reset_index()
  concern_data = concern_data.rename(columns = {0: 'Recency_Customers'})
  concern_data = pd.merge(concern_data, product_data, how = 'inner', on = ['Marital_Status', 'Education', 
                                                                           'Income_Class', 'Age_Class',
                                                                           'Kidhome', 'Teenhome'])
  concern_data['Recency_Percentage'] = 100 * concern_data['Recency_Customers'] / concern_data['Customers']
  concern_data = concern_data.sort_values(by = ['Potential_Percentage', 'Recency_Percentage'], ascending = False).reset_index(drop = True)
  
  return concern_data

## Wines

In [11]:
# Potential Wines Customer
temp = get_potential_customer(data, 'Wines_Cluster', 2)

wines = get_potential_percentage(data, temp, min_customer)
wines

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                1          3             3          0        0         0   
1                1          3             3          2        0         1   
2                1          3             3          1        0         0   
3                0          2             3          2        0         0   
4                1          3             3          3        0         0   
5                0          3             2          2        0         1   
6                0          3             3          2        0         0   
7                0          2             3          0        0         0   
8                1          1             3          2        0         1   
9                1          1             3          3        0         0   
10               1          2             3          2        0         0   
11               0          3             2          3        0         1   
12               0          1             3          1        0         0   
13               1          2             3          3        0         0   
14               1          3             3          2        0         0   
15               0          1             3          2        0         0   
16               1          1             3          3        0         1   
17               1          2             2          1        0         1   
18               1          3             2          2        0         1   
19               1          1             3          1        0         0   
20               0          1             3          3        0         0   
21               1          1             3          0        0         0   
22               1          1             3          2        0         0   
23               1          2             2          3        0         1   
24               0          2             2          2        0         1   
25               1          3             2          3        0         1   
26               1          1             3          1        0         1   
27               0          1             2          2        0         1   
28               1          3             2          1        0         1   
29               1          2             2          2        0         1   
30               0          1             2          3        0         1   
31               1          1             2          1        0         1   
32               0          1             3          2        0         1   
33               1          1             2          2        0         1   
34               1          1             2          3        0         1   
35               1          1             2          2        0         0   
36               0          1             3          0        0         0   
37               1          2             2          3        0         0   

    Customers  Potential_Customers  Potential_Percentage  
0          11                    9             81.818182  
1          10                    8             80.000000  
2          15                    9             60.000000  
3          10                    6             60.000000  
4          16                    9             56.250000  
5          14                    7             50.000000  
6          14                    7             50.000000  
7          11                    5             45.454545  
8          22                   10             45.454545  
9          32                   14             43.750000  
10         23                    9             39.130435  
11         11                    4             36.363636  
12         18                    6             33.333333  
13         15                    5             33.333333  
14         15                    5             33.333333  
15         18                    6             33.333333

Only locate the characteristics in which the potential customer population are higher than certain threshold.
Here, we will set the threshold as **60%** because that is a reasonable number to say if certain characteristics are what we seek in our potential customers.

In [12]:
threshold = 60
top_wines = wines.loc[wines['Potential_Percentage'] >= threshold].reset_index(drop = True)
top_wines.head(10)

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0               1          3             3          0        0         0   
1               1          3             3          2        0         1   
2               1          3             3          1        0         0   
3               0          2             3          2        0         0   

   Customers  Potential_Customers  Potential_Percentage  
0         11                    9             81.818182  
1         10                    8             80.000000  
2         15                    9             60.000000  
3         10                    6             60.000000

From above, we can see that the potential wine buyers are most likely to have these criterias:

First segment
1.   Is **married**.
2.   Possess a **doctoral degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   Is **younger than 40**.
5.   **Do not** have any kid nor teen at home.

Second segment
1.   Is **married**.
2.   Possess a **doctoral degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   Around **50 - 65** years old.
5.   **Do not** have any kide and **have** 1 teen at home.

Third segment
1.   Is **married**.
2.   Possess a **doctoral degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   **Around 40 - 50** years old.
5.   **Do not** have any kid nor teen at home.

Now, we will check if these mentioned above are active customers.

In [13]:
wines_potential = data.loc[((data['Marital_Status'] == 1) &
                            (data['Education'] == 3) &
                            (data['Income_Class'] == 3) &
                            ((data['Age_Class'] == 0) | (data['Age_Class'] == 1)) &
                            (data['Kidhome'] == 0) &
                            (data['Teenhome'] == 0)) |
                           
                           ((data['Marital_Status'] == 1) &
                            (data['Education'] == 3) &
                            (data['Income_Class'] == 3) &
                            (data['Age_Class'] == 2) &
                            (data['Kidhome'] == 0) &
                            (data['Teenhome'] == 1))] \
                    .groupby(['Marital_Status', 'Education', 
                              'Income_Class', 'Age_Class', 
                              'Kidhome', 'Teenhome']) \
                    .value_counts().reset_index()

display_rfm(wines_potential, 'Wines_Cluster')

Recency_Cluster
0    38.889
1    41.667
2    19.444
Name: value, dtype: float64

Frequency_Cluster
0     8.333
1    58.333
2    33.333
Name: value, dtype: float64

Wines_Cluster
0     5.556
1    22.222
2    72.222
Name: value, dtype: float64

Notice that the precentage of **inactive potential customers** are **quite high**, even though they are the ones who contribute the most to our GDP. This is concerning.

In [14]:
# Look for overall potential customers (almost) churn distribution
mask = rfm[(rfm['Recency_Cluster'] == 0) & (rfm['Wines_Cluster'] == 2)]['ID'].values
concern_wines = get_concern(data, wines, mask)
concern_wines

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                1          3             3          0        0         0   
1                1          3             3          2        0         1   
2                1          3             3          1        0         0   
3                0          2             3          2        0         0   
4                1          3             3          3        0         0   
5                0          3             3          2        0         0   
6                0          3             2          2        0         1   
7                0          2             3          0        0         0   
8                1          1             3          2        0         1   
9                1          1             3          3        0         0   
10               1          2             3          2        0         0   
11               0          3             2          3        0         1   
12               1          2             3          3        0         0   
13               0          1             3          2        0         0   
14               1          3             3          2        0         0   
15               1          2             2          1        0         1   
16               1          3             2          2        0         1   
17               1          1             3          1        0         0   
18               0          1             3          3        0         0   
19               1          1             3          0        0         0   
20               1          1             3          2        0         0   
21               1          2             2          3        0         1   
22               1          3             2          3        0         1   
23               1          1             3          1        0         1   
24               0          1             2          2        0         1   
25               1          3             2          1        0         1   
26               0          1             2          3        0         1   
27               1          1             2          2        0         1   
28               1          1             2          3        0         1   
29               0          1             3          0        0         0   
30               1          2             2          3        0         0   

    Recency_Customers  Customers  Potential_Customers  Potential_Percentage  \
0                   2         11                    9             81.818182   
1                   2         10                    8             80.000000   
2                   6         15                    9             60.000000   
3                   3         10                    6             60.000000   
4                   2         16                    9             56.250000   
5                   4         14                    7             50.000000   
6                   2         14                    7             50.000000   
7                   3         11                    5             45.454545   
8                   4         22                   10             45.454545   
9                   2         32                   14             43.750000   
10                  1         23                    9             39.130435   
11                  3         11                    4             36.363636   
12                  2         15                    5             33.333333   
13                  2         18                    6             33.333333   
14                  1         15                    5             33.333333   
15                  2         10                    3             30.000000   
16                  1         34                   10             29.411765   
17                  6         26                    7             26.923077   
18                  3         23      

Notice that there is **no centralized characteristics** in which certain customer segments are inactive. Also, notice that our potential customers' churn percentages are **relatively low**, which are **<=20%**, **except** for the customers who are **40-50 years old**. Thus, there are 2 options we can do here:

1. Stage a **discount / campaign** which can attract them to buy wines again.
2. **Divert** our **potential customers** to the fourth segments, which is exactly the same as the third, except that they are **>65** years old. This is because their **potential** rate is quite high, **56.25%**, with a **churn** rate of only **12.5%**.


## Fruits

In [15]:
# Potential Wines Customer
temp = get_potential_customer(data, 'Fruits_Cluster', 2)

fruits = get_potential_percentage(data, temp, min_customer)
fruits

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          1             3          1        0         0   
1                0          1             3          2        0         0   
2                1          1             2          2        0         0   
3                1          1             3          0        0         0   
4                1          2             3          2        0         0   
5                0          1             3          0        0         0   
6                1          1             3          3        0         1   
7                0          3             3          2        0         0   
8                1          1             3          3        0         0   
9                1          1             3          1        0         0   
10               0          1             3          2        0         1   
11               1          1             3          2        0         0   
12               1          1             3          1        0         1   
13               1          2             3          3        0         0   
14               1          3             3          2        0         0   
15               0          2             3          0        0         0   
16               1          1             2          1        0         1   
17               1          1             2          2        0         1   
18               0          1             2          1        0         1   
19               1          2             2          2        0         1   
20               0          2             3          2        0         0   
21               1          3             3          2        0         1   
22               1          3             3          0        0         0   
23               1          2             2          3        0         0   
24               0          1             2          3        0         1   
25               0          1             2          2        0         1   
26               0          2             2          2        0         1   
27               1          3             3          1        0         0   
28               1          3             3          3        0         0   
29               1          1             2          3        0         1   
30               1          1             3          2        0         1   
31               0          1             3          3        0         0   

    Customers  Potential_Customers  Potential_Percentage  
0          18                    9             50.000000  
1          18                    9             50.000000  
2          10                    5             50.000000  
3          20                    9             45.000000  
4          23                   10             43.478261  
5          23                    8             34.782609  
6          10                    3             30.000000  
7          14                    4             28.571429  
8          32                    9             28.125000  
9          26                    7             26.923077  
10         15                    4             26.666667  
11         45                   12             26.666667  
12         16                    4             25.000000  
13         15                    3             20.000000  
14         15                    3             20.000000  
15         11                    2             18.181818  
16         26                    4             15.384615  
17         55                    8             14.545455  
18         14                    2             14.285714  
19         18                    2             11.111111  
20         10                    1             10.000000  
21         10                    1             10.000000  
22         11                    1              9.090909  
23         12                    1            

Only locate the characteristics in which the potential customer population are higher than certain threshold.
Here, we will set the threshold as **60%** because that is a reasonable number to say if certain characteristics are what we seek in our potential customers.

But then, notice that there is **no characteristic** which **exceedes** that threshold. Thus, we will lower the threshold to **50%**.

In [16]:
threshold = 50
top_fruits = fruits.loc[fruits['Potential_Percentage'] >= threshold].reset_index(drop = True)
top_fruits.head(10)

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0               0          1             3          1        0         0   
1               0          1             3          2        0         0   
2               1          1             2          2        0         0   

   Customers  Potential_Customers  Potential_Percentage  
0         18                    9                  50.0  
1         18                    9                  50.0  
2         10                    5                  50.0

From above, we can see that the potential fruits buyers are most likely to have these criterias:

First segment
1.   Is **single**.
2.   Possess a **graduate degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   Around **40 - 65** years old.
5.   **Do not** have any kid nor teen at home.

Second segment
1.   Is **married**.
2.   Possess a **graduate degree**.
3.   Have an **upper-middle income** (USD 51K++ Annual income).
4.   Around **50 - 65** years old.
5.   **Do not** have any kid nor teen at home.

Now, we will check if these mentioned above are active customers.

In [17]:
fruits_potential = data.loc[((data['Marital_Status'] == 0) &
                             (data['Education'] == 1) &
                             (data['Income_Class'] == 3) &
                             ((data['Age_Class'] == 1) | (data['Age_Class'] == 2)) &
                             (data['Kidhome'] == 0) &
                             (data['Teenhome'] == 0))|
                            
                            ((data['Marital_Status'] == 1) &
                             (data['Education'] == 1) &
                             (data['Income_Class'] == 2) &
                             (data['Age_Class'] == 2) &
                             (data['Kidhome'] == 0) &
                             (data['Teenhome'] == 0))] \
                    .groupby(['Marital_Status', 'Education', 
                              'Income_Class', 'Age_Class', 
                              'Kidhome', 'Teenhome']) \
                    .value_counts().reset_index()

display_rfm(fruits_potential, 'Fruits_Cluster')

Recency_Cluster
0    26.087
1    19.565
2    54.348
Name: value, dtype: float64

Frequency_Cluster
0    23.913
1    54.348
2    21.739
Name: value, dtype: float64

Fruits_Cluster
0    17.391
1    32.609
2    50.000
Name: value, dtype: float64

Notice that the precentage of **inactive potential customers** are terribly **high**, even though they are the ones who contribute the most to our GDP. This is concerning.

In [18]:
# Look for overall potential customers (almost) churn distribution
mask = rfm[(rfm['Recency_Cluster'] == 0) & (rfm['Fruits_Cluster'] == 2)]['ID'].values
concern_fruits = get_concern(data, fruits, mask)
concern_fruits

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          1             3          2        0         0   
1                0          1             3          1        0         0   
2                1          1             2          2        0         0   
3                1          1             3          0        0         0   
4                1          2             3          2        0         0   
5                0          1             3          0        0         0   
6                0          3             3          2        0         0   
7                1          1             3          3        0         0   
8                1          1             3          1        0         0   
9                1          1             3          2        0         0   
10               1          2             3          3        0         0   
11               1          3             3          2        0         0   
12               0          2             3          0        0         0   
13               1          1             2          1        0         1   
14               1          1             2          2        0         1   
15               0          1             2          1        0         1   
16               1          2             2          2        0         1   
17               1          3             3          2        0         1   
18               0          2             2          2        0         1   
19               1          3             3          1        0         0   
20               1          1             3          2        0         1   

    Recency_Customers  Customers  Potential_Customers  Potential_Percentage  \
0                   4         18                    9             50.000000   
1                   2         18                    9             50.000000   
2                   1         10                    5             50.000000   
3                   4         20                    9             45.000000   
4                   2         23                   10             43.478261   
5                   2         23                    8             34.782609   
6                   1         14                    4             28.571429   
7                   3         32                    9             28.125000   
8                   5         26                    7             26.923077   
9                   8         45                   12             26.666667   
10                  2         15                    3             20.000000   
11                  1         15                    3             20.000000   
12                  2         11                    2             18.181818   
13                  1         26                    4             15.384615   
14                  4         55                    8             14.545455   
15                  2         14                    2             14.285714   
16                  2         18                    2             11.111111   
17                  1         10                    1             10.000000   
18                  1         14                    1              7.142857   
19                  1         15                    1              6.666667   
20                  1         22                    1              4.545455   

    Recency_Percentage  
0            22.222222  
1            11.111111  
2            10.000000  
3            20.000000  
4             8.695652  
5             8.695652  
6             7.142857  
7             9.375000  
8            19.230769  
9            17.777778  
10           13.333333  
11            6.666667  
12           18.181818  
13            3.846154  
14            7.272727  
15           14.285714  
16           11.111111  
17           10.000000  
18            7.142857  
19            6.666667  
20            4.545455

Notice that there is **no centralized characteristics** in which certain customer segments are inactive. Also, notice that our potential customers' churn percentages are **relatively low**, which are **<=20%**, **except** for the customers who are **50-65 years old**. Since it only **exceeds** the threshold by **2%** and we **can not divert** our attention to another customer segment (since they have low potential rate), the only option we have is to stage a **discount / campaign** which can attract them to buy fruits again.

## Meat

In [19]:
# Potential Meat Customer
temp = get_potential_customer(data, 'Meat_Cluster', 2)

meat = get_potential_percentage(data, temp, min_customer)
meat

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          2             3          2        0         0   
1                0          2             3          0        0         0   
2                0          1             3          2        0         0   
3                1          1             3          1        0         0   
4                1          1             3          3        0         0   
5                0          1             3          0        0         0   
6                1          1             3          2        0         0   
7                1          3             3          2        0         0   
8                1          2             3          3        0         0   
9                1          1             3          0        0         0   
10               0          3             3          2        0         0   
11               1          3             3          3        0         0   
12               0          1             3          1        0         0   
13               1          3             3          1        0         0   
14               1          2             3          2        0         0   
15               1          3             3          0        0         0   
16               0          1             3          3        0         0   
17               1          1             3          1        0         1   
18               1          2             2          3        0         0   
19               1          1             2          2        0         0   
20               1          3             3          2        0         1   
21               1          1             3          2        0         1   
22               1          1             0          1        1         0   

    Customers  Potential_Customers  Potential_Percentage  
0          10                    8             80.000000  
1          11                    7             63.636364  
2          18                   11             61.111111  
3          26                   14             53.846154  
4          32                   16             50.000000  
5          23                   11             47.826087  
6          45                   21             46.666667  
7          15                    7             46.666667  
8          15                    7             46.666667  
9          20                    9             45.000000  
10         14                    6             42.857143  
11         16                    6             37.500000  
12         18                    6             33.333333  
13         15                    5             33.333333  
14         23                    7             30.434783  
15         11                    3             27.272727  
16         23                    6             26.086957  
17         16                    3             18.750000  
18         12                    2             16.666667  
19         10                    1             10.000000  
20         10                    1             10.000000  
21         22                    1              4.545455  
22         49                    1              2.040816

Only locate the characteristics in which the potential customer population are higher than certain threshold.
Here, we will set the threshold as **60%** because that is a reasonable number to say if certain characteristics are what we seek in our potential customers.

In [20]:
threshold = 60
top_meat = meat.loc[meat['Potential_Percentage'] >= threshold].reset_index(drop = True)
top_meat.head(10)

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0               0          2             3          2        0         0   
1               0          2             3          0        0         0   
2               0          1             3          2        0         0   

   Customers  Potential_Customers  Potential_Percentage  
0         10                    8             80.000000  
1         11                    7             63.636364  
2         18                   11             61.111111

From above, we can see that the potential meat buyers are most likely to have these criterias:

First segment
1.   Is **single**.
2.   Possess a **master degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   Is **younger than 40** or around **50 - 65** years old.
5.   **Do not** have any kid nor teen at home.

Second segment
1.   Is **single**.
2.   Possess a **graduate degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   Around **50 - 65** years old.
5.   **Do not** have any kid nor teen at home.

Now, we will check if these mentioned above are active customers.

In [21]:
meat_potential = data.loc[((data['Marital_Status'] == 0) &
                           (data['Education'] == 2) &
                           (data['Income_Class'] == 3) &
                           (((data['Age_Class'] == 0)) | (data['Age_Class'] == 2)) &
                           (data['Kidhome'] == 0) &
                           (data['Teenhome'] <= 0)) |
                          
                          ((data['Marital_Status'] == 0) &
                           (data['Education'] == 1) &
                           (data['Income_Class'] == 3) &
                           (data['Age_Class'] == 2) &
                           (data['Kidhome'] == 0) &
                           (data['Teenhome'] <= 0))] \
                    .groupby(['Marital_Status', 'Education', 
                              'Income_Class', 'Age_Class', 
                              'Kidhome', 'Teenhome']) \
                    .value_counts().reset_index()

display_rfm(meat_potential, 'Meat_Cluster')

Recency_Cluster
0    38.462
1    25.641
2    35.897
Name: value, dtype: float64

Frequency_Cluster
0    15.385
1    69.231
2    15.385
Name: value, dtype: float64

Meat_Cluster
0     2.564
1    30.769
2    66.667
Name: value, dtype: float64

Notice that the precentage of **inactive potential customers** are terribly **high**, even though they are the ones who contribute the most to our GDP. This is concerning.

In [22]:
# Look for overall potential customers (almost) churn distribution
mask = rfm[(rfm['Recency_Cluster'] == 0) & (rfm['Meat_Cluster'] == 2)]['ID'].values
concern_meat = get_concern(data, meat, mask)
concern_meat

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          2             3          2        0         0   
1                0          2             3          0        0         0   
2                0          1             3          2        0         0   
3                1          1             3          1        0         0   
4                1          1             3          3        0         0   
5                0          1             3          0        0         0   
6                1          1             3          2        0         0   
7                1          2             3          3        0         0   
8                1          3             3          2        0         0   
9                1          1             3          0        0         0   
10               0          3             3          2        0         0   
11               1          3             3          1        0         0   
12               0          1             3          1        0         0   
13               1          2             3          2        0         0   
14               1          3             3          0        0         0   
15               1          1             3          1        0         1   
16               1          1             2          2        0         0   
17               1          3             3          2        0         1   
18               1          1             3          2        0         1   

    Recency_Customers  Customers  Potential_Customers  Potential_Percentage  \
0                   3         10                    8             80.000000   
1                   2         11                    7             63.636364   
2                   3         18                   11             61.111111   
3                   8         26                   14             53.846154   
4                   5         32                   16             50.000000   
5                   3         23                   11             47.826087   
6                   9         45                   21             46.666667   
7                   3         15                    7             46.666667   
8                   1         15                    7             46.666667   
9                   4         20                    9             45.000000   
10                  2         14                    6             42.857143   
11                  3         15                    5             33.333333   
12                  1         18                    6             33.333333   
13                  2         23                    7             30.434783   
14                  1         11                    3             27.272727   
15                  1         16                    3             18.750000   
16                  1         10                    1             10.000000   
17                  1         10                    1             10.000000   
18                  1         22                    1              4.545455   

    Recency_Percentage  
0            30.000000  
1            18.181818  
2            16.666667  
3            30.769231  
4            15.625000  
5            13.043478  
6            20.000000  
7            20.000000  
8             6.666667  
9            20.000000  
10           14.285714  
11           20.000000  
12            5.555556  
13            8.695652  
14            9.090909  
15            6.250000  
16           10.000000  
17           10.000000  
18            4.545455

Notice that there is **no centralized characteristics** in which certain customer segments are inactive. Also, notice that our potential customers' churn percentages are **relatively low**, which are **<=20%**, **except** for the customers who are **50-65 years old**. Since it **exceeds** the threshold by quite a number, which is **10%** and they are basically our main customers for this product, we have to **work hard** and make sure that they **won't** leave us. And so, one option we have is to stage a **discount / campaign** which can attract them to buy meat again.

## Fish

In [23]:
# Potential Fish Customer
temp = get_potential_customer(data, 'Fish_Cluster', 2)

fish = get_potential_percentage(data, temp, min_customer)
fish

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          1             3          1        0         0   
1                1          1             3          3        0         1   
2                0          2             3          2        0         0   
3                1          1             3          2        0         0   
4                0          1             3          0        0         0   
5                0          1             3          2        0         0   
6                1          1             3          3        0         0   
7                1          1             3          1        0         1   
8                0          3             3          2        0         0   
9                1          2             3          2        0         0   
10               1          1             3          1        0         0   
11               1          1             3          0        0         0   
12               1          2             3          3        0         0   
13               0          1             3          2        0         1   
14               1          3             3          3        0         0   
15               0          1             3          3        0         0   
16               1          2             2          2        0         1   
17               1          2             2          3        0         0   
18               0          2             2          2        0         1   
19               0          1             2          1        0         1   
20               1          3             3          2        0         0   
21               1          1             2          1        0         1   
22               1          2             2          3        0         1   
23               1          3             3          2        0         1   
24               1          1             2          2        0         0   
25               0          1             0          2        0         0   
26               0          2             3          0        0         0   
27               1          3             3          1        0         0   
28               1          1             2          3        0         1   
29               1          1             3          2        0         1   
30               0          1             2          2        0         1   
31               1          1             2          2        0         1   

    Customers  Potential_Customers  Potential_Percentage  
0          18                    9             50.000000  
1          10                    4             40.000000  
2          10                    4             40.000000  
3          45                   18             40.000000  
4          23                    9             39.130435  
5          18                    7             38.888889  
6          32                   12             37.500000  
7          16                    6             37.500000  
8          14                    5             35.714286  
9          23                    8             34.782609  
10         26                    9             34.615385  
11         20                    6             30.000000  
12         15                    4             26.666667  
13         15                    4             26.666667  
14         16                    4             25.000000  
15         23                    4             17.391304  
16         18                    3             16.666667  
17         12                    2             16.666667  
18         14                    2             14.285714  
19         14                    2             14.285714  
20         15                    2             13.333333  
21         26                    3             11.538462  
22         18                    2             11.111111  
23         10                    1            

Only locate the characteristics in which the potential customer population are higher than certain threshold.
Here, we will set the threshold as **60%** because that is a reasonable number to say if certain characteristics are what we seek in our potential customers.

But, notice that there is **no customer segment** which exceeds that threshold. Thus, we will **customize** the threshold to **50%** in this case.

In [24]:
threshold = 50
top_fish = fish.loc[fish['Potential_Percentage'] >= threshold].reset_index(drop = True)
top_fish.head(10)

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0               0          1             3          1        0         0   

   Customers  Potential_Customers  Potential_Percentage  
0         18                    9                  50.0

From above, we can see that the potential fish buyers are most likely to have these criterias:

First segment
1.   Is **single**.
2.   Possess a **graduate degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   Around **40 - 50** years old.
5.   **Do not** have any kid nor teen at home.

Now, we will check if these mentioned above are active customers.

In [25]:
fish_potential = data.loc[((data['Marital_Status'] == 0) &
                           (data['Education'] == 1) &
                           (data['Income_Class'] == 3) &
                           (data['Age_Class'] == 1) &
                           (data['Kidhome'] == 0) &
                           (data['Teenhome'] == 0))] \
                    .groupby(['Marital_Status', 'Education', 
                              'Income_Class', 'Age_Class', 
                              'Kidhome', 'Teenhome']) \
                    .value_counts().reset_index()

display_rfm(fish_potential, 'Fish_Cluster')

Recency_Cluster
0    16.667
1    27.778
2    55.556
Name: value, dtype: float64

Frequency_Cluster
0    22.222
1    61.111
2    16.667
Name: value, dtype: float64

Fish_Cluster
0    27.778
1    22.222
2    50.000
Name: value, dtype: float64

Notice that the precentage of **inactive potential customers** are terribly **high**, even though they are the ones who contribute the most to our GDP. This is concerning.

In [26]:
# Look for overall potential customers (almost) churn distribution
mask = rfm[(rfm['Recency_Cluster'] == 0) & (rfm['Fish_Cluster'] == 2)]['ID'].values
concern_fish = get_concern(data, fish, mask)
concern_fish

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                1          1             3          2        0         0   
1                1          1             3          3        0         1   
2                0          1             3          0        0         0   
3                0          1             3          2        0         0   
4                1          1             3          3        0         0   
5                1          1             3          1        0         1   
6                0          3             3          2        0         0   
7                1          2             3          2        0         0   
8                1          1             3          1        0         0   
9                1          1             3          0        0         0   
10               0          1             3          2        0         1   
11               1          2             3          3        0         0   
12               1          2             2          2        0         1   
13               1          2             2          3        0         0   
14               0          1             2          1        0         1   
15               0          2             2          2        0         1   
16               1          3             3          2        0         0   
17               1          1             2          1        0         1   
18               1          3             3          2        0         1   
19               1          3             3          1        0         0   
20               1          1             3          2        0         1   
21               0          1             2          2        0         1   

    Recency_Customers  Customers  Potential_Customers  Potential_Percentage  \
0                   7         45                   18             40.000000   
1                   1         10                    4             40.000000   
2                   4         23                    9             39.130435   
3                   1         18                    7             38.888889   
4                   5         32                   12             37.500000   
5                   1         16                    6             37.500000   
6                   2         14                    5             35.714286   
7                   2         23                    8             34.782609   
8                   5         26                    9             34.615385   
9                   3         20                    6             30.000000   
10                  1         15                    4             26.666667   
11                  1         15                    4             26.666667   
12                  2         18                    3             16.666667   
13                  1         12                    2             16.666667   
14                  2         14                    2             14.285714   
15                  2         14                    2             14.285714   
16                  1         15                    2             13.333333   
17                  1         26                    3             11.538462   
18                  1         10                    1             10.000000   
19                  1         15                    1              6.666667   
20                  1         22                    1              4.545455   
21                  1         27                    1              3.703704   

    Recency_Percentage  
0            15.555556  
1            10.000000  
2            17.391304  
3             5.555556  
4            15.625000  
5             6.250000  
6            14.285714  
7             8.695652  
8            19.230769  
9            15.000000  
10            6.666667  
11            6.666667  
12           11.111111  
13            8.333333  
14           14.285714  
15           1

Notice that there is **no centralized characteristics** in which certain customer segments are inactive. Thus, we will try to **generalize** the **marketing strategies** to bring back these types of customers, as well as targeting new potential customers who match these criterias.

In addition to that, notice that the table above shows that our best potential customer here does not have a customer which has a 0 Recency_Cluster, which means they are all (almost) active customers.

## Sweets

In [27]:
# Potential Sweets Customer
temp = get_potential_customer(data, 'Sweets_Cluster', 2)

sweets = get_potential_percentage(data, temp, min_customer)
sweets

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          1             3          2        0         0   
1                0          2             3          2        0         0   
2                0          1             3          2        0         1   
3                0          2             3          0        0         0   
4                1          1             3          0        0         0   
5                0          1             3          1        0         0   
6                1          1             3          3        0         0   
7                1          1             3          1        0         0   
8                1          2             3          3        0         0   
9                1          1             3          1        0         1   
10               1          2             3          2        0         0   
11               1          1             3          2        0         1   
12               0          1             3          3        0         0   
13               1          1             3          2        0         0   
14               1          1             3          3        0         1   
15               1          1             2          1        0         1   
16               1          3             3          0        0         0   
17               1          2             2          2        0         1   
18               0          2             2          2        0         1   
19               0          3             3          2        0         0   
20               1          3             3          1        0         0   
21               1          3             3          2        0         0   
22               0          1             3          0        0         0   
23               1          1             2          2        0         1   
24               1          2             2          3        0         0   
25               0          1             2          3        0         1   
26               0          1             2          2        0         1   
27               1          2             1          1        1         0   

    Customers  Potential_Customers  Potential_Percentage  
0          18                   10             55.555556  
1          10                    5             50.000000  
2          15                    7             46.666667  
3          11                    5             45.454545  
4          20                    8             40.000000  
5          18                    7             38.888889  
6          32                   12             37.500000  
7          26                    9             34.615385  
8          15                    5             33.333333  
9          16                    5             31.250000  
10         23                    6             26.086957  
11         22                    5             22.727273  
12         23                    5             21.739130  
13         45                    9             20.000000  
14         10                    2             20.000000  
15         26                    5             19.230769  
16         11                    2             18.181818  
17         18                    3             16.666667  
18         14                    2             14.285714  
19         14                    2             14.285714  
20         15                    2             13.333333  
21         15                    2             13.333333  
22         23                    3             13.043478  
23         55                    5              9.090909  
24         12                    1              8.333333  
25         13                    1              7.692308  
26         27                    2              7.407407  
27         15                    1              6.666667

Only locate the characteristics in which the potential customer population are higher than certain threshold.
Here, we will set the threshold as **60%** because that is a reasonable number to say if certain characteristics are what we seek in our potential customers.

But, notice that there is **no segment** which exceeds that threshold. Thus, we will **customize** the threshold to **50%** in this case.

In [28]:
threshold = 50
top_sweets = sweets.loc[sweets['Potential_Percentage'] >= threshold].reset_index(drop = True)
top_sweets.head(10)

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0               0          1             3          2        0         0   
1               0          2             3          2        0         0   

   Customers  Potential_Customers  Potential_Percentage  
0         18                   10             55.555556  
1         10                    5             50.000000

From above, we can see that the potential sweets buyers are most likely to have these criterias:

First segment
1.   Marital status does not matter, but it is **more likely** if they are married.
2.   Possess a **graduate degree**.
3.   Have a **high income** (USD 68K++ Annual income).
4.   **Do not** have any kid nor teen at home.   

Now, we will check if these mentioned above are active customers.

In [29]:
sweets_potential = data.loc[(data['Education'] == 1) &
                            (data['Income_Class'] == 3) &
                            (data['Kidhome'] == 0) &
                            (data['Teenhome'] == 0)] \
                    .groupby(['Marital_Status', 'Education', 
                              'Income_Class', 'Age_Class', 
                              'Kidhome', 'Teenhome']) \
                    .value_counts().reset_index()

display_rfm(sweets_potential, 'Sweets_Cluster')

Recency_Cluster
0    36.585
1    28.780
2    34.634
Name: value, dtype: float64

Frequency_Cluster
0    10.732
1    72.683
2    16.585
Name: value, dtype: float64

Sweets_Cluster
0    33.659
1    35.610
2    30.732
Name: value, dtype: float64

Notice that the precentage of **inactive potential customers** are terribly **high**, even though they are the ones who contribute the most to our GDP. This is concerning.

In [30]:
# Look for overall potential customers (almost) churn distribution
mask = rfm[(rfm['Recency_Cluster'] == 0) & (rfm['Sweets_Cluster'] == 2)]['ID'].values
concern_sweets = get_concern(data, sweets, mask)
concern_sweets

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          1             3          2        0         0   
1                0          2             3          2        0         0   
2                0          1             3          2        0         1   
3                0          2             3          0        0         0   
4                1          1             3          0        0         0   
5                0          1             3          1        0         0   
6                1          1             3          3        0         0   
7                1          1             3          1        0         0   
8                1          2             3          3        0         0   
9                1          1             3          1        0         1   
10               1          2             3          2        0         0   
11               1          1             3          2        0         1   
12               1          1             3          2        0         0   
13               1          1             2          1        0         1   
14               1          2             2          2        0         1   
15               0          2             2          2        0         1   
16               1          3             3          2        0         0   
17               0          1             3          0        0         0   
18               1          1             2          2        0         1   

    Recency_Customers  Customers  Potential_Customers  Potential_Percentage  \
0                   4         18                   10             55.555556   
1                   3         10                    5             50.000000   
2                   1         15                    7             46.666667   
3                   3         11                    5             45.454545   
4                   3         20                    8             40.000000   
5                   2         18                    7             38.888889   
6                   3         32                   12             37.500000   
7                   3         26                    9             34.615385   
8                   2         15                    5             33.333333   
9                   2         16                    5             31.250000   
10                  1         23                    6             26.086957   
11                  1         22                    5             22.727273   
12                  4         45                    9             20.000000   
13                  1         26                    5             19.230769   
14                  2         18                    3             16.666667   
15                  1         14                    2             14.285714   
16                  2         15                    2             13.333333   
17                  1         23                    3             13.043478   
18                  1         55                    5              9.090909   

    Recency_Percentage  
0            22.222222  
1            30.000000  
2             6.666667  
3            27.272727  
4            15.000000  
5            11.111111  
6             9.375000  
7            11.538462  
8            13.333333  
9            12.500000  
10            4.347826  
11            4.545455  
12            8.888889  
13            3.846154  
14           11.111111  
15            7.142857  
16           13.333333  
17            4.347826  
18            1.818182

Notice that there is **no centralized characteristics** in which certain customer segments are inactive. Thus, we will try to **generalize** the **marketing strategies** to bring back these types of customers, as well as targeting new potential customers who match these criterias.

Also, notice that our potential buyers all have high churn rate, which are **>20%**. And since the **other segments** have considerably **low potential** rate, it is recommended to **divert** our focus to make them **keep buying** our product instead of promoting to other customer segments.

## Gold

In [31]:
# Potential Gold Customer
temp = get_potential_customer(data, 'Gold_Cluster', 2)

gold = get_potential_percentage(data, temp, min_customer)
gold

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          2             2          2        0         1   
1                1          1             3          3        0         1   
2                1          1             3          0        0         0   
3                0          1             2          1        0         1   
4                0          1             3          1        0         0   
5                0          1             3          3        0         0   
6                0          2             3          2        0         0   
7                1          1             3          2        0         0   
8                1          1             3          3        0         0   
9                1          1             3          2        0         1   
10               1          3             3          1        0         0   
11               1          2             3          3        0         0   
12               0          1             3          0        0         0   
13               1          1             2          2        0         1   
14               1          2             2          3        0         0   
15               1          2             2          3        0         1   
16               0          1             2          2        0         1   
17               0          3             2          2        0         1   
18               1          1             2          3        0         1   
19               1          3             3          2        0         1   
20               1          1             1          1        0         1   
21               1          1             2          1        0         1   
22               0          1             2          3        0         1   
23               0          1             3          2        0         1   
24               1          2             3          2        0         0   
25               1          1             3          1        0         1   
26               1          3             3          3        0         0   
27               1          1             3          1        0         0   
28               1          1             1          0        1         0   
29               1          2             2          1        0         1   
30               0          1             1          2        0         1   
31               0          3             2          3        0         1   
32               0          2             3          0        0         0   
33               1          1             2          2        1         1   
34               1          1             1          3        1         1   
35               1          2             1          2        0         1   
36               0          3             3          2        0         0   
37               1          2             1          1        1         0   
38               1          3             3          2        0         0   
39               1          3             2          2        0         1   
40               1          1             0          0        1         0   
41               0          1             3          2        0         0   
42               1          3             2          3        0         1   

    Customers  Potential_Customers  Potential_Percentage  
0          14                    6             42.857143  
1          10                    4             40.000000  
2          20                    8             40.000000  
3          14                    5             35.714286  
4          18                    6             33.333333  
5          23                    7             30.434783  
6          10                    3             30.000000  
7          45                   13             28.888889  
8          32                    9             28.125000  
9          22            

Only locate the characteristics in which the potential customer population are higher than certain threshold.
Here, we will set the threshold as **60%** because that is a reasonable number to say if certain characteristics are what we seek in our potential customers.

But, notice that there is **no customer segment** which exceeds that threshold. And since the **highest** potential rate is merely **42%**, we can't say that they are our potential customer. So, we have 2 option here:
1. **Do not** spend any $ to market gold-related product to save money.
2. **Bet** our money into the **"best" customer segment**, even thuogh the success rate is only 42%.

In case we choose the 2nd option, then we will get the following results.

In [32]:
threshold = 60
top_gold = gold.iloc[0].reset_index(drop = True)
top_gold.head(10)

0     0.000000
1     2.000000
2     2.000000
3     2.000000
4     0.000000
5     1.000000
6    14.000000
7     6.000000
8    42.857143
Name: 0, dtype: float64

From above, we can see that the potential gold buyers are most likely to have these criterias:

First segment
1.   Is **married**.
2.   Possess a **graduate degree**.
3.   Have a **high-income** (USD 68K++ Annual income).
4.   **At least** in their **40s**.
5.   **Do not** have any kid and **at most** 1 teen at home.   

Now, we will check if these mentioned above are active customers.

In [33]:
gold_potential = data.loc[((data['Marital_Status'] == 0) &
                           (data['Education'] == 2) &
                           (data['Income_Class'] == 2) &
                           (data['Age_Class'] == 2) &
                           (data['Kidhome'] == 0) &
                           (data['Teenhome'] == 1))] \
                    .groupby(['Marital_Status', 'Education', 
                              'Income_Class', 'Age_Class', 
                              'Kidhome', 'Teenhome']) \
                    .value_counts().reset_index()

display_rfm(gold_potential, 'Gold_Cluster')

Recency_Cluster
0    28.571
1    35.714
2    35.714
Name: value, dtype: float64

Frequency_Cluster
1    21.429
2    78.571
Name: value, dtype: float64

Gold_Cluster
0    35.714
1    21.429
2    42.857
Name: value, dtype: float64

Notice that the precentage of **inactive potential customers** are terribly **high**, even though they are the ones who contribute the most to our GDP. This is concerning.

In [34]:
# Look for overall potential customers (almost) churn distribution
mask = rfm[(rfm['Recency_Cluster'] == 0) & (rfm['Gold_Cluster'] == 2)]['ID'].values
concern_gold = get_concern(data, gold, mask)
concern_gold

Marital_Status  Education  Income_Class  Age_Class  Kidhome  Teenhome  \
0                0          2             2          2        0         1   
1                1          1             3          0        0         0   
2                1          1             3          3        0         1   
3                0          1             2          1        0         1   
4                0          1             3          1        0         0   
5                0          1             3          3        0         0   
6                0          2             3          2        0         0   
7                1          1             3          2        0         0   
8                1          1             3          3        0         0   
9                1          1             3          2        0         1   
10               1          3             3          1        0         0   
11               0          1             3          0        0         0   
12               1          1             2          2        0         1   
13               1          2             2          3        0         0   
14               0          3             2          2        0         1   
15               1          1             2          3        0         1   
16               1          3             3          2        0         1   
17               0          1             2          3        0         1   
18               1          1             2          1        0         1   
19               0          1             3          2        0         1   
20               1          2             3          2        0         0   
21               1          1             3          1        0         1   
22               1          1             3          1        0         0   
23               0          1             1          2        0         1   
24               1          1             1          3        1         1   
25               0          3             3          2        0         0   
26               1          2             1          2        0         1   
27               1          2             1          1        1         0   
28               1          3             2          2        0         1   
29               0          1             3          2        0         0   

    Recency_Customers  Customers  Potential_Customers  Potential_Percentage  \
0                   1         14                    6             42.857143   
1                   5         20                    8             40.000000   
2                   1         10                    4             40.000000   
3                   3         14                    5             35.714286   
4                   1         18                    6             33.333333   
5                   1         23                    7             30.434783   
6                   1         10                    3             30.000000   
7                   4         45                   13             28.888889   
8                   3         32                    9             28.125000   
9                   2         22                    6             27.272727   
10                  1         15                    4             26.666667   
11                  2         23                    6             26.086957   
12                  5         55                   14             25.454545   
13                  2         12                    3             25.000000   
14                  2         14                    3             21.428571   
15                  3         20                    4             20.000000   
16                  1         10                    2             20.000000   
17                  2         13                    2             15.384615   
18                  1         26                    4             15.384615   
19                  1         15    

Notice that there is **no centralized characteristics** in which certain customer segments are inactive. Thus, we will try to **generalize** the **marketing strategies** to bring back these types of customers, as well as targeting new potential customers who match these criterias.

In addition to that, our **"best" customer segment** also has **very low churn** rate, which is only **7%**, and it is obviously a good thing. We should work on **promoting** the **product** more to **this customer segment**.